### ISSUE TITLE
Queues are not always re-created properly when nodes in a clustered RabbitMQ are restarted

### ISSUE THREAD

**Issue by brianloss** [spring-projects/spring-amqp#818]
I am using spring-amqp (2.0.5.RELEASE) via spring-cloud-bus and spring-cloud-stream with a clustered RabbitMQ setup. Spring-cloud-bus creates an auto-delete, exclusive queue for each service I run. Sometimes when I restart one of the RabbitMQ nodes, one or more of my services will fail to re-create the associated queues that went away. Instead, the service will be stuck looping forever with an exception in BlockingQueueConsumer.attemptPassiveDeclarations.

If service1 connects to node rabbit1 and creates its auto-delete queue which ends up being owned by rabbit1, then everything is fine. If I restart rabbit1 then service1 detects that the connection went away and it re-establishes the connection and also queue(s). However, I also see the scenario where service1 has a connection to, say rabbit2, but the auto-delete queue is owned by, say rabbit3. If I were to restart rabbit3, then the queue goes away. However, service1 still has a good connection to rabbit2. It detects that the consumer for the auto-delete queue has gone away and attempts to restart that, but it then does the passive queue declaration which fails, and then it just keeps retrying forever.

Is there an option that will cause queue(s) to be re-created if the passive declaration fails?

**Comment by garyrussell** [comment]
I am not sure how that's possible; usually, auto-delete queues are also exclusive, which means they are "owned" by the connection that creates them.

What is the use case?

**Comment by garyrussell** [comment]
There is, however, an issue with auto-delete queues (used by Spring Bus) when the underlying RabbitMQ connection factory has `autoRecoveryEnabled=true`.

Normally spring amqp disables it, but if you supply a rabbitmq connection factory with it set, it will cause this problem.

If you are running on PCF, spring-cloud-connectors prior to 2.0.3 leaves this true; it is fixed in 2.0.3.

**Comment by garyrussell** [comment]
One work around:

```java
@Bean
public ApplicationRunner runner(CachingConnectionFactory cf) {
	cf.getRabbitConnectionFactory().setAutomaticRecoveryEnabled(false);
	return args -> { };
}
```

**Comment by garyrussell** [comment]
You should check with the RabbitMQ guys on the rabbitmq-users google group but it doesn't make sense to me to mirror auto delete exclusive queues.

**Comment by brianloss** [comment]
Thanks for the quick replies! Sorry for not being clear. The queues created by spring-cloud-bus are exclusive auto-delete, and I made no attempt to mirror those (I don't believe it's possible with RabbitMQ, according to their [docs](http://www.rabbitmq.com/ha.html#master-migration-data-locality)). One of our microservices also creates its own queues that are not auto-delete (and are durable). Those queues are mirrored by policy. They all live in the same rabbitmq cluster.

I'll see if I can replicate in a smaller example such that I'm able to post some code.

**Comment by garyrussell** [comment]
> one or more of my services will fail to re-create the associated queues that went away.

> One of our microservices also creates its own queues that are not auto-delete (and are durable)

So, why did they "went away", and why do they need to be redeclared?

**Comment by brianloss** [comment]
The auto-delete queues (the ones created for me by spring-cloud-bus) disappear upon restart of the rabbitmq cluster node that owned them. Since, in some cases, the service that created the auto-delete queue was somehow connected to a different rabbitmq cluster node, the connection never failed and the consumer for the queue in the creating service sits and spins forever waiting for the queue to re-appear. The durable queues we create that are mirrored survive a rabbitmq node restart.

**Comment by garyrussell** [comment]
>Since, in some cases, the service that created the auto-delete queue was somehow connected to a different rabbitmq cluster node

That's the part I don't understand; I don't see how it's possible for the queue to be on a different node to the one with which the connection is established, since the queue is created on that connection and is "owned" by that connection.

**Comment by brianloss** [comment]
I agree it's weird, but that's what seems to be happening. I'm working on reproducing in a smaller example such that I can post some code/logs/etc.

**Comment by brianloss** [comment]
Haven't had any luck reproducing this on a small example, but I can definitely confirm that some queues are created on a rabbit node that is not the node of a client connection. I turned up logging to trace for org.springframework.amqp.rabbit, and was able to verify that all calls relating to the creation of one example springCloudBus exclusive queue matched the connection, which was to rabbit-0. In the logs for rabbit-0, I see the connection listed. However, in the logs for rabbit-1, I see the log for the creation of the exclusive queue and rabbit-1 ends up being the owner.

I'll ask in the rabbitmq users group if this is something that is supposed to happen. In the mean time, is there any work-around I can implement to re-create the queue when the consumer dies?

**Comment by garyrussell** [comment]
Very strange.

I just tested this with a bus app..

```java
	@Autowired
	private AmqpAdmin admin;

	@EventListener
	public void events(ListenerContainerConsumerFailedEvent event) {
		SimpleMessageListenerContainer container = (SimpleMessageListenerContainer) event.getSource();
		String queueName = container.getQueueNames()[0];
		if (queueName.startsWith("springCloudBus")) {
			this.admin.declareQueue(new Queue(queueName, false, true, true, null));
		}
	}
```

and it worked ok. The queue declaration is idempotent so it won't matter if the queue was already re-declared.

**Comment by brianloss** [comment]
I found [this](https://groups.google.com/forum/#!msg/rabbitmq-users/YbikYfl_0y4/XI7j1hyLCAAJ;context-place=forum/rabbitmq-users) thread on the rabbitmq users lists which describes exactly my problem. My rabbitmq cluster was using the `min-master` queue master locator.

According to the [docs](http://www.rabbitmq.com/ha.html#master-migration-data-locality), it is possible to set a queue master locator per queue as an argument when declaring the queue. I would think possibly spring-rabbit needs to do this when creating an exclusive queue. Or, perhaps it could be documented that using spring-rabbit with a cluster and a queue master locator other than client-local is not supported? Third option would be to recreate queues whenever the consumer/producer connection fails vs relying on total connection failure. I'd guess in that case you'd only want to do that after some timeout to allow for a durable queue that disappeared (because its master was on a node other than the connection node) to come back when the node is restarted.

**Comment by brianloss** [comment]
I forgot to mention that once I changed the queue master locator to `client-local`, I stopped seeing the behavior of exclusive queues getting mastered on nodes other than the one holding the connection.

**Comment by garyrussell** [comment]
Interesting; even with `min-master` I couldn't get it to create a queue on a different node when I was testing earlier - it always created them on the connected node.

**Comment by garyrussell** [comment]
https://jira.spring.io/browse/AMQP-834

**Comment by brianloss** [comment]
Maybe it's just luck? Or if you're connecting to the rabbit nodes in a round-robin fashion then you'd likely get the same behavior for `client-local` and `min-master`. In the k8s environment, it's using DNS to contact a rabbit node, and that must randomize access to the nodes a bit more.

Anyway, thanks for the help and the quick replies. Much appreciated.

**Comment by dsteegen** [comment]
Hi people,

I am experiencing the same problem as described by the thread starter. I have a minimal setup where I use Spring AMQP 2.1.6.RELEASE together with Spring Cloud Bus 2.1.1.RELEASE.
The application creates an anonymous queue in RabbitMQ but fails to re-create the missing queue when it was deleted (e.g. network outage).
It keeps logging every 5 seconds the following message

```
2019-05-20 14:42:55.477  WARN 10711 --- [pplication-2-99] o.s.a.r.listener.BlockingQueueConsumer   : Failed to declare queue: springCloudBus.test-application-2
2019-05-20 14:42:55.478  WARN 10711 --- [pplication-2-99] o.s.a.r.listener.BlockingQueueConsumer   : Queue declaration failed; retries left=1

org.springframework.amqp.rabbit.listener.BlockingQueueConsumer$DeclarationException: Failed to declare queue(s):[springCloudBus.test-application-2]
	at org.springframework.amqp.rabbit.listener.BlockingQueueConsumer.attemptPassiveDeclarations(BlockingQueueConsumer.java:710) [spring-rabbit-2.1.6.RELEASE.jar:2.1.6.RELEASE]
	at org.springframework.amqp.rabbit.listener.BlockingQueueConsumer.passiveDeclarations(BlockingQueueConsumer.java:594) [spring-rabbit-2.1.6.RELEASE.jar:2.1.6.RELEASE]
	at org.springframework.amqp.rabbit.listener.BlockingQueueConsumer.start(BlockingQueueConsumer.java:581) [spring-rabbit-2.1.6.RELEASE.jar:2.1.6.RELEASE]
	at org.springframework.amqp.rabbit.listener.SimpleMessageListenerContainer$AsyncMessageProcessingConsumer.initialize(SimpleMessageListenerContainer.java:1196) [spring-rabbit-2.1.6.RELEASE.jar:2.1.6.RELEASE]
	at org.springframework.amqp.rabbit.listener.SimpleMessageListenerContainer$AsyncMessageProcessingConsumer.run(SimpleMessageListenerContainer.java:1041) [spring-rabbit-2.1.6.RELEASE.jar:2.1.6.RELEASE]
	at java.lang.Thread.run(Thread.java:748) [na:1.8.0_201]
Caused by: java.io.IOException: null
	at com.rabbitmq.client.impl.AMQChannel.wrap(AMQChannel.java:126) ~[amqp-client-5.4.3.jar:5.4.3]
	at com.rabbitmq.client.impl.AMQChannel.wrap(AMQChannel.java:122) ~[amqp-client-5.4.3.jar:5.4.3]
	at com.rabbitmq.client.impl.AMQChannel.exnWrappingRpc(AMQChannel.java:144) ~[amqp-client-5.4.3.jar:5.4.3]
	at com.rabbitmq.client.impl.ChannelN.queueDeclarePassive(ChannelN.java:1006) ~[amqp-client-5.4.3.jar:5.4.3]
	at com.rabbitmq.client.impl.ChannelN.queueDeclarePassive(ChannelN.java:52) ~[amqp-client-5.4.3.jar:5.4.3]
	at sun.reflect.GeneratedMethodAccessor149.invoke(Unknown Source) ~[na:na]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_201]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[na:1.8.0_201]
	at org.springframework.amqp.rabbit.connection.CachingConnectionFactory$CachedChannelInvocationHandler.invoke(CachingConnectionFactory.java:1140) ~[spring-rabbit-2.1.6.RELEASE.jar:2.1.6.RELEASE]
	at com.sun.proxy.$Proxy129.queueDeclarePassive(Unknown Source) ~[na:na]
	at org.springframework.amqp.rabbit.listener.BlockingQueueConsumer.attemptPassiveDeclarations(BlockingQueueConsumer.java:689) [spring-rabbit-2.1.6.RELEASE.jar:2.1.6.RELEASE]
	... 5 common frames omitted
Caused by: com.rabbitmq.client.ShutdownSignalException: channel error; protocol method: #method<channel.close>(reply-code=404, reply-text=NOT_FOUND - no queue 'springCloudBus.test-application-2' in vhost '/', class-id=50, method-id=10)
	at com.rabbitmq.utility.ValueOrException.getValue(ValueOrException.java:66) ~[amqp-client-5.4.3.jar:5.4.3]
	at com.rabbitmq.utility.BlockingValueOrException.uninterruptibleGetValue(BlockingValueOrException.java:36) ~[amqp-client-5.4.3.jar:5.4.3]
	at com.rabbitmq.client.impl.AMQChannel$BlockingRpcContinuation.getReply(AMQChannel.java:494) ~[amqp-client-5.4.3.jar:5.4.3]
	at com.rabbitmq.client.impl.AMQChannel.privateRpc(AMQChannel.java:288) ~[amqp-client-5.4.3.jar:5.4.3]
	at com.rabbitmq.client.impl.AMQChannel.exnWrappingRpc(AMQChannel.java:138) ~[amqp-client-5.4.3.jar:5.4.3]
	... 13 common frames omitted
Caused by: com.rabbitmq.client.ShutdownSignalException: channel error; protocol method: #method<channel.close>(reply-code=404, reply-text=NOT_FOUND - no queue 'springCloudBus.test-application-2' in vhost '/', class-id=50, method-id=10)
	at com.rabbitmq.client.impl.ChannelN.asyncShutdown(ChannelN.java:516) ~[amqp-client-5.4.3.jar:5.4.3]
	at com.rabbitmq.client.impl.ChannelN.processAsync(ChannelN.java:346) ~[amqp-client-5.4.3.jar:5.4.3]
	at com.rabbitmq.client.impl.AMQChannel.handleCompleteInboundCommand(AMQChannel.java:178) ~[amqp-client-5.4.3.jar:5.4.3]
	at com.rabbitmq.client.impl.AMQChannel.handleFrame(AMQChannel.java:111) ~[amqp-client-5.4.3.jar:5.4.3]
	at com.rabbitmq.client.impl.AMQConnection.readFrame(AMQConnection.java:670) ~[amqp-client-5.4.3.jar:5.4.3]
	at com.rabbitmq.client.impl.AMQConnection.access$300(AMQConnection.java:48) ~[amqp-client-5.4.3.jar:5.4.3]
	at com.rabbitmq.client.impl.AMQConnection$MainLoop.run(AMQConnection.java:597) ~[amqp-client-5.4.3.jar:5.4.3]
	... 1 common frames omitted
```

I am using a non clustered setup for RabbitMQ in my local environment although the issue is the same on our production environment where we have 3 Rabbits running in a cluster.

Is there a way to get around this issue? I tried setting the property `missingQueuesFatal` to `true` but that will just stop the `SimpleMessageListenerContainer` after a few seconds.

**Comment by garyrussell** [comment]
@dsteegen Please don't comment on old, closed, issues.

We are not aware of any issues in this area; can you look in the logs? You should see this...

>2019-05-20 10:22:46.918  INFO 16729 --- [OSpMfnyEdfkdQ-7] o.s.a.r.c.CachingConnectionFactory       : Created new connection: rabbitConnectionFactory#e72dba7:11/SimpleConnection@1bbbf38c [delegate=amqp://guest@127.0.0.1:5672/, localPort= 65480]
2019-05-20 10:22:46.921  INFO 16729 --- [OSpMfnyEdfkdQ-7] o.s.amqp.rabbit.core.RabbitAdmin         : Auto-declaring a non-durable, auto-delete, or exclusive Queue (springCloudBus.anonymous.CmIerG8RSOSpMfnyEdfkdQ) durable:false, auto-delete:true, exclusive:true. It will be redeclared if the broker stops and is restarted while the connection factory is alive, but all messages will be lost.
2019-05-20 10:23:38.788 ERROR 16729 --- [ 127.0.0.1:5672] o.s.a.r.c.CachingConnectionFactory       : Channel shutdown: connection error; protocol method: #method<connection.close>(reply-code=320, reply-text=CONNECTION_FORCED - Closed via management plugin, class-id=0, method-id=0)
2019-05-20 10:23:39.113  INFO 16729 --- [OSpMfnyEdfkdQ-7] o.s.a.r.l.SimpleMessageListenerContainer : Restarting Consumer@6a16ad31: tags=[[amq.ctag-TsqOhRfkMsFGM1n8btcUXw]], channel=Cached Rabbit Channel: AMQChannel(amqp://guest@127.0.0.1:5672/,1), conn: Proxy@46fe0445 Shared Rabbit Connection: SimpleConnection@1bbbf38c [delegate=amqp://guest@127.0.0.1:5672/, localPort= 65480], acknowledgeMode=AUTO local queue size=0
2019-05-20 10:23:39.114  INFO 16729 --- [OSpMfnyEdfkdQ-8] o.s.a.r.c.CachingConnectionFactory       : Attempting to connect to: [localhost:5672]
2019-05-20 10:23:39.119  INFO 16729 --- [OSpMfnyEdfkdQ-8] o.s.a.r.c.CachingConnectionFactory       : Created new connection: rabbitConnectionFactory#e72dba7:12/SimpleConnection@70c28bbb [delegate=amqp://guest@127.0.0.1:5672/, localPort= 65502]
**2019-05-20 10:23:39.120  INFO 16729 --- [OSpMfnyEdfkdQ-8] o.s.amqp.rabbit.core.RabbitAdmin         : Auto-declaring a non-durable, auto-delete, or exclusive Queue (springCloudBus.anonymous.CmIerG8RSOSpMfnyEdfkdQ) durable:false, auto-delete:true, exclusive:true. It will be redeclared if the broker stops and is restarted while the connection factory is alive, but all messages will be lost.**

**Comment by dsteegen** [comment]
@garyrussell 

My apologies, I will create a new issue next time instead of replying to a closed ticket.

But to answer your question:
I only see that logging message when I stop the RabbitMQ docker instance and restarting it afterwards. When I remove the queue directly via the RabbitMQ admin console, I expected the same behaviour but instead I see the error message as described in my previous comment.
Note that I had to configure the `group` property for the anonymous queues as this will make the queue non-exclusive and thus also deletable via the admin console.

Let me know if you want me to create a new ticket for this. I will then also add a sample project that allows you to reproduce the problem.

**Comment by garyrussell** [comment]
>When I remove the queue directly via the RabbitMQ admin console, I expected the same behaviour 

You should not expect the same behavior - manually deleting the queue will not trigger a re-declaration of the queue.

The `RabbitAdmin` adds a `ConnectionListener` to the connection factory.

Only when a new connection is created is the admin triggered to re-create the queue. When the connection is forced closed, the container goes into recovery mode; when the connection is opened again, we re-create the queue.

When you delete a queue, the consumer is canceled.

    /**
     * Called when the consumer is cancelled for reasons <i>other than</i> by a call to
     * {@link Channel#basicCancel}. For example, the queue has been deleted.
     * See {@link #handleCancelOk} for notification of consumer
     * cancellation due to {@link Channel#basicCancel}.
     * @param consumerTag the <i>consumer tag</i> associated with the consumer
     * @throws IOException
     */
    void handleCancel(String consumerTag) throws IOException;

**Comment by dsteegen** [comment]
Thanks for clarifying! I was actually just trying to mimic the issue we sometimes have on our production environment where clustering is setup. II thought that manually deleting a queue in a non clustered environment would end up in the same situation as it spawns the same exception.

Nevertheless, after digging deeper into the problem I came to the conclusion that I am experiencing the same problem as the thread starter. So I thought that it would be fixed with the changes linked to this ticket but that is not the case since the `AnonymousQueue` class that sets the `x-queue-master-locator` property to `client-local` is not used by `spring-cloud-stream-binder-rabbit`.
Therefor I opened a new ticket on that project.

(again, sorry for hijacking this closed issue :))
